app/services/katello/event_daemon.rb

```diff
@@ -1,60 +1,91 @@
       end
 
-      if $PROGRAM_NAME.match(/spring/)
-        Rails.logger.info("Spring is running in the environment; not starting Katello Event Daemon")
-        return false
+      def settings
+        SETTINGS[:katello][:event_daemon]
       end
 
-      if Rails.env.production? && $PROGRAM_NAME.match(/dynflow/)
-        Rails.logger.info("This appears to be the Dynflow process; not starting Katello Event Daemon")
-        return false
+      def pid
+        return unless pid_file && File.exist?(pid_file)
+
+        File.open(pid_file) { |f| f.read.to_i }
       end
 
-      if started?
-        Rails.logger.info("Katello Event Daemon already running; not attempting startup")
-        return false
+      def pid_file
+        pid_dir.join('katello_event_daemon.pid')
       end
 
-      !::Foreman.in_rake? && !Rails.env.test? && SETTINGS[:katello][:event_daemon_enabled]
-    end
+      def pid_dir
+        Rails.root.join('tmp', 'pids')
+      end
 
-    def self.start
-      if runnable?
-        services.values.each(&:run)
+      def lock_file
+        Rails.root.join('tmp', 'katello_event_daemon.lock')
+      end
 
-        at_exit do
-          stop
-        end
+      def write_pid_file
+        return unless pid_file
 
-        @pid = Process.pid
+        FileUtils.mkdir_p(pid_dir)
+        File.open(pid_file, 'w') { |f| f.puts Process.pid }
       end
-    end
 
-    def self.stop
-      # in puma clustered mode, we should only stop the daemon
-      # from the parent process (where it actually is running), not forked workers
-      return unless @pid == Process.pid
+      def stop
+        return unless pid == Process.pid
+        services.values.each(&:close)
+        File.unlink(pid_file) if pid_file && File.exist?(pid_file)
+      end
 
-      services.values.each(&:close)
+      def start
+        return unless runnable?
 
-      @pid = nil
-    end
+        lockfile = File.open(lock_file, 'r')
+        begin
+          lockfile.flock(File::LOCK_EX)
+          return if started? # ensure it wasn't started while we waited for the lock
 
-    def self.started?
-      @pid.present?
-    end
+          start_services
+          write_pid_file
+
+          at_exit do
+            stop
+          end
 
-    def self.services
-      {
-        candlepin_events: ::Katello::CandlepinEventListener,
-        katello_events: ::Katello::EventMonitor::PollerThread
-      }
+          Rails.logger.info("Katello event daemon started process=#{Process.pid}")
+        ensure
+          lockfile.flock(File::LOCK_UN)
+        end
+      end
+
+      def started?
+        Process.kill(0, pid)
+        true
+      rescue Errno::ESRCH, TypeError # process no longer exists or we had no PID cached
+        false
+      end
+
+      def start_services
+        services.values.each(&:run)
+      end
+
+      def runnable?
+        # avoid accessing the disk on each request
+        @cache.fetch('katello_event_daemon_runnable', expires_in: 1.minute) do
```

Comment:
This adds a nice performance benefit over a lot of requests (100k in these results)

```
=> [#<Benchmark::Tms:0x0000000013d78828 @label="runnable_cached", @real=5.545715187967289, @cstime=0.0, @cutime=0.0, @stime=0.1517280000000003, @utime=1.3842180000000042, @total=1.5359460000000045>]


=> [#<Benchmark::Tms:0x000000000999ae68 @label="runnable_uncached", @real=79.99794052902143, @cstime=0.0, @cutime=0.0, @stime=1.329981, @utime=14.650019, @total=15.98>]```